securedrop_client/gui/actions.py

```diff
@@ -126,3 +132,136 @@ def _on_confirmation_dialog_accepted(self) -> None:
                 return
             self.controller.delete_conversation(self.source)
             self._state.remove_conversation_files(id)
+
+
+class PrintConversation(QAction):
+    """Use this action to print a transcript of the messages, replies, etc.
+
+    The transcript includes references to any attached files.
+    """
+
+    printer_start_requested = pyqtSignal()
+    printer_job_enqueued = pyqtSignal(str, str)
+
+    _SUPPORT_FOR_CONCURRENT_PRINTING_JOBS_ENABLED = False
+
+    def __init__(
+        self,
+        source: Source,
+        parent: QMenu,
+        controller: Controller,
+        printer: Printer,
+        confirmation_dialog: Callable[[Printer, str], QDialog],
+        error_dialog: Callable[[str, str], QDialog],
+    ) -> None:
+        title = _("Print Conversation")
+        super().__init__(title, parent)
+
+        self._controller = controller
+        self._source = source
+        self._printer = printer
+        self._create_error_dialog = error_dialog
+        self._create_confirmation_dialog = confirmation_dialog
+        self._printing_job_id = self._source.journalist_designation
+        self._printing_job_failure_notification_in_progress = False
+
+        self.setShortcut(Qt.CTRL + Qt.Key_P)
+
+        self._printer.job_failed.connect(self._on_printing_job_failed)
+        self._printer.job_done.connect(self._on_printing_job_done)
+        self.triggered.connect(self.trigger)
+
+        self._printer.start_and_watch_on(self.printer_start_requested)
+        self._printer.enqueue_job_on(self.printer_job_enqueued)
+        self._printer.stop_watching_on(self.printer_job_enqueued)
+        self._printer.stop_watching_on(self._printer.job_failed)
```

Comment:
This is not test-covered. We must make sure that all exits remain handled. (I made sure of that manually, but that should be automatically tested.)